include/aspect/stokes_matrix_free.h

```diff
@@ -0,0 +1,652 @@
+                  (viscosity_x_2(cell,q)*velocity.get_symmetric_gradient(q),q);
+                }
+              velocity.integrate (false,true);
+
+              diagonal[i] = velocity.begin_dof_values()[i];
+            }
+
+          for (unsigned int i=0; i<velocity.dofs_per_cell; ++i)
+            velocity.begin_dof_values()[i] = diagonal[i];
+          velocity.distribute_local_to_global (dst);
+        }
+    }
+  }
+
+
+
+  template<int dim>
+  class StokesMatrixFreeHandler
+  {
+    public:
+      /**
+                 * Initialize this class, allowing it to read in
+                 * relevant parameters as well as giving it a reference to the
+                 * Simulator that owns it, since it needs to make fairly extensive
+                 * changes to the internals of the simulator.
+                 */
+      StokesMatrixFreeHandler(Simulator<dim> &, ParameterHandler &prm);
+
+      /**
+                 * Destructor
+                 */
+      ~StokesMatrixFreeHandler();
+
+      /**
+                 * Solves the Stokes linear system matrix-free. This is called
+                 * by Simulator<dim>::solve_stokes().
+                 */
+      std::pair<double,double> solve();
+
+      /**
+                 * Allocates and sets up the members of the StokesMatrixFreeHandler. This
+                 * is called by Simulator<dim>::setup_dofs()
+                 */
+      void setup_dofs();
+
+      /**
+             * Evalute the MaterialModel to query for the viscosity on the active cells,
+             * project this viscosity to the multigrid hierarchy, and cache the information
+             * for later usage.
+             */
+      void evaluate_viscosity();
+
+      /**
+       * Get the workload imbalance of the distribution
+       * of the level hierarchy
+       */
+      double get_workload_imbalance();
+
+      /**
+             * Add correction to system RHS for non-zero boundary condition
+             */
+      void correct_stokes_rhs();
+
+      /**
+                 * Declare parameters. (No actual parameters at the moment)
+                 */
+      static
+      void declare_parameters (ParameterHandler &prm);
+
+      /**
+                 * Parse parameters. (No actual parameters at the moment)
+                 */
+      void parse_parameters (ParameterHandler &prm);
+
+
+    private:
+
+      Simulator<dim> &sim;
+
+      // TODO: velocity degree not only 2, Choosing quadrature degree?
+      typedef MatrixFreeStokesOperators::StokesOperator<dim,2,double> StokesMatrixType;
+      typedef MatrixFreeStokesOperators::MassMatrixOperator<dim,1,double> MassMatrixType;
+      typedef MatrixFreeStokesOperators::ABlockOperator<dim,2,double> ABlockMatrixType;
+
+      DoFHandler<dim> dof_handler_v;
+      DoFHandler<dim> dof_handler_p;
+      DoFHandler<dim> dof_handler_projection;
+
+      FESystem<dim> stokes_fe;
+      FESystem<dim> fe_v;
+      FESystem<dim> fe_p;
+      const FESystem<dim> fe_projection;
```

Comment:
why const?

Reply:
No reason, copied and pasted from somewhere. Should all these be const or not? We never change them.